src/libopensc/cwa-dnie.c

```diff
@@ -724,17 +728,12 @@ static int dnie_transmit_apdu_internal(sc_card_t * card, sc_apdu_t * apdu)
 
 		size_t e_txlen = 0;
 		size_t index = 0;
-		sc_apdu_t *e_apdu = NULL;
-		u8 *e_tx = NULL;
+		sc_apdu_t e_apdu;
+		u8 e_tx[2*SC_MAX_APDU_BUFFER_SIZE];
 
 		/* envelope needed */
 		sc_log(card->ctx, "envelope tx required: lc:%d", apdu->lc);
 
-		e_apdu = calloc(1, sizeof(sc_apdu_t));	/* enveloped apdu */
-		e_tx = calloc(7 + apdu->datalen, sizeof(u8));	/* enveloped data */
```

Comment:
The main goal of this fix is to avoid the memory leaks. I consider that changing from dynamic to static allocation is a minor performance improvement. Once that you have highlighted the potential problems of this change, I prefer to stick to dynamic allocation for e_tx and free the memory in all the correct places, which doesn't seem that complex in this case. It seems more future proof, and performance is a minor problem now. When there are no memory leaks at all and there are no functional problems, maybe we have time to make additional improvements. So, unless you object, I will make these changes. Thank you again for spotting this.